pandas/core/frame.py

```diff
@@ -3967,8 +3967,15 @@ def _set_value(
         """
         try:
             if takeable:
-                series = self._ixs(col, axis=1)
-                series._set_value(index, value, takeable=True)
+                if isinstance(self._mgr, ArrayManager):
+                    # with CoW, we can't use intermediate series
+                    # with takeable=True, we know that index is positional and
+                    # not a generic hashable label
+                    index = cast(int, index)
+                    self._mgr.column_setitem(col, index, value)
+                else:
+                    series = self._ixs(col, axis=1)
+                    series._set_value(index, value, takeable=True)
```

Comment:
if we go down this path, we should make a BlockManager.column_setitem

Reply:
I would have to look into, but this might not be that straightforward, since the current path also deals with updating the cache etc